### PR TITLE
Fix prepared param lost when there is only one null param value for proxy

### LINKED ISSUE #38612: Proxy may lose null parameter when executing prepared statement
### Which version of ShardingSphere did you use?
5.5.4-SNAPSHOT 5f7e8abe7079815f79515f33154122701aae2920

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy

### Expected behavior
The goal is for the proxy (MySQL protocol) to behave consistently with native MySQL.

### Actual behavior
When executing PreparedStatement and specifying the only param value as null, native MySQL executes successfully, while proxy reports an error.

```java
@SneakyThrows
private static void queryWithPreparedStatement(final HikariDataSource dataSource) {
    try (Connection connection = dataSource.getConnection();
         PreparedStatement statement = connection.prepareStatement("SELECT * FROM t_single where name = ?")) {
        log.info("jdbcUrl: {}", dataSource.getJdbcUrl());
        statement.setString(1, null);
//            statement.setString(1, "name");
        statement.execute();
        ResultSet resultSet =  statement.getResultSet();
        while (resultSet.next()) {
            log.info("{}, {}, {}", resultSet.getInt(1), resultSet.getString(2), resultSet.getString(3));
        }
    }
}
```

- Native MySQL

Image

- Proxy

Image

### Reason analyze (If you can)
When only one parameter is NULL, the MySQL client driver (such as MySQL Connector/J) employs the following optimization strategy:

- Short form optimization: When all parameters are detected to be NULL, the driver may choose to send the data packet in short form, without sending parameter type information, because:
  - NULL values ​​do not require specific data type information for decoding
  - Reduce network transmission overhead
  - Assume the server can handle the data based on type information obtained during the pre-compilation stage
- Missing parameter type information:
  - In the MySQLComStmtExecutePacket constructor, if newParametersBoundFlag is not PARAMETER_TYPE_EXIST, newParameterTypes is set to Collections.emptyList()
  - This results in preparedStatement.getParameterTypes() also being empty, thus preventing the setting of params.

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
- Create logic database
```sql
create database sharding_db;

use sharding_db;
```

- Register storage unit
```sql
REGISTER STORAGE UNIT ds_0 (
    URL="jdbc:mysql://127.0.0.1:3306/demo_ds_0?serverTimezone=Asia/Shanghai&characterEncoding=UTF-8&useSSL=false&allowPublicKeyRetrieval=true",
    USER="root",
    PASSWORD="123456",
    PROPERTIES("maximumPoolSize"=10,"minimumIdle"=1)
);
```

- Create table
```sql
DROP TABLE IF EXISTS t_single;

CREATE TABLE `t_single` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(32) NOT NULL,
  `pwd` varchar(50) NOT NULL,
  `addr` varchar(50) NOT NULL,
  PRIMARY KEY (`id`)
);
```

- Insert data
```sql
INSERT INTO t_single (name, pwd, addr) values ('name', 'pwd', 'addr');

SELECT * FROM t_single;
```

- Execute query from JDBC client

### PR DESCRIPTION
**Pull request by RaigorJiang**
Fixes #38612.
 
#### Changes proposed in this pull request:
  - fix the MySQL Proxy prepared statement execution path when Connector/J sends `COM_STMT_EXECUTE` in short form and does not include parameter type
  metadata
  - add a fallback in `MySQLComStmtExecuteExecutor` to rebuild parameter types from the prepared statement parameter count and null bitmap when
  parameter types are missing
  - ensure null parameters are preserved instead of being lost during parameter binding, so Proxy behaves consistently with native MySQL
  - preserve the original behavior for requests that already provide parameter types through `newParametersBoundFlag` or cached prepared statement
  metadata
  - add regression tests to cover single-null, partial-null, all-null, and non-null prepared statement parameter scenarios

### CODE REVIEW

File: proxy/frontend/dialect/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/execute/MySQLComStmtExecuteExecutor.java [hunk 1]
```diff
@@ -85,6 +87,30 @@ public Collection<DatabasePacket> execute() throws SQLException {
...
+        if (!originalTypes.isEmpty()) {
+            return originalTypes;
+        }
+        List<MySQLPreparedStatementParameterType> result = new ArrayList<>(expectedParamCount);
+        for (int i = 0; i < expectedParamCount; i++) {
+            if (null != packet.getNullBitmap() && packet.getNullBitmap().isNullParameter(i)) {
+                result.add(new MySQLPreparedStatementParameterType(MySQLBinaryColumnType.NULL, 0));
+            } else {
+                result.add(new MySQLPreparedStatementParameterType(MySQLBinaryColumnType.VARCHAR, 0));
+            }
```
[the remark this answers is not in the document]
Comment (RaigorJiang) [reply proxy/frontend/dialect/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/execute/MySQLComStmtExecuteExecutor.java]: Yes, I will tighten the getParameterTypes method to only handle cases where all parameters are NULL, and will not provide a VARCHAR fallback.

File: proxy/frontend/dialect/mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/execute/MySQLComStmtExecuteExecutorTest.java [hunk 2]
```diff
@@ -208,4 +211,100 @@ void assertExecutePreparedCommit() throws SQLException {
...
+        MySQLNullBitmap nullBitmap = mock(MySQLNullBitmap.class);
+        when(nullBitmap.isNullParameter(0)).thenReturn(true);
+        when(packet.getNullBitmap()).thenReturn(nullBitmap);
+        MySQLComStmtExecuteExecutor executor = new MySQLComStmtExecuteExecutor(packet, connectionSession);
+        when(proxyBackendHandler.execute()).thenReturn(new UpdateResponseHeader(selectStatement));
+        when(ProxyBackendHandlerFactory.newInstance(any(MySQLDatabaseType.class), any(QueryContext.class), eq(connectionSession), anyBoolean())).thenReturn(proxyBackendHandler);
+        Iterator<DatabasePacket> actual = executor.execute().iterator();
+        assertThat(actual.next(), isA(MySQLOKPacket.class));
+        assertFalse(actual.hasNext());
+    }
```
[the remark this answers is not in the document]
Comment (RaigorJiang) [reply proxy/frontend/dialect/mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/execute/MySQLComStmtExecuteExecutorTest.java:257]: Refactored

File: proxy/frontend/dialect/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/execute/MySQLComStmtExecuteExecutor.java [hunk 3]
```diff
@@ -85,6 +88,38 @@ public Collection<DatabasePacket> execute() throws SQLException {
         return responseHeader instanceof QueryResponseHeader ? processQuery((QueryResponseHeader) responseHeader) : processUpdate((UpdateResponseHeader) responseHeader);
     }
     
+    private List<MySQLPreparedStatementParameterType> getParameterTypes(final MySQLServerPreparedStatement preparedStatement) {
+        List<MySQLPreparedStatementParameterType> parameterTypes = preparedStatement.getParameterTypes();
+        if (!parameterTypes.isEmpty()) {
+            return parameterTypes;
+        }
+        List<MySQLPreparedStatementParameterType> originalTypes = packet.getNewParameterTypes();
+        if (!originalTypes.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (RaigorJiang) [reply proxy/frontend/dialect/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/execute/MySQLComStmtExecuteExecutor.java]: OK

**Review by taojintianxia — APPROVED** [review]
